datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
-        if re.search(pattern='[^a-z0-9_\-.\\/\ ]', string=remote_dir, flags=re.ASCII | re.IGNORECASE):
-            self.annex.error(f"Invalid character in directory name of {remote_file}."
-                             f"Valid characters are a-Z, 0-9, '_', '-', '.', '\\', '/' and ' ' (white space).")
-
-        datafile = Datafile()
-        datafile.set({'filename': remote_file,
-                      'directoryLabel': remote_dir,
-                      'label': os.path.basename(remote_file)})
-        self.transfer_store(key=remote_file, local_file=local_file, datafile=datafile)
 
-    def transfer_retrieve(self, key, file):
+        # TODO: Properly test whether presentexport should consider all dataset
+        # versions or just latest. Problem: To retrieve "old" keys, all versions
+        # would be relevant, but for upload a reexport of the same file may be
+        # the correct thing (present in version 1, removed in 2, about to be
+        # present in 3 again).
+        known_state = self.annex.getstate(key)
+        if known_state:  # empty string otherwise
+            return int(known_state) in self.files_latest.keys()
+        else:
+            return mangle_directory_names(remote_file) in \
+                   [p for p in self.files_latest.values()]
+
+    def _upload_file(self, datafile, key, local_file, remote_file):
+        """helper for both transfer-store methods"""
+        # If the remote path already exists, we need to replace rather than
+        # upload the file, since otherwise dataverse would rename the file on
+        # its end. However, this only concerns the latest version of the dataset
+        # (which is what we are pushing into)!
+        existing_id = [i for i, p in self.files_latest.items()
+                       if p == remote_file]
+        if existing_id:
+            self.message(f"Replacing {remote_file} ...", type='debug')
+            # Note: docstring of identifier appears to be incorrect. It's the id
+            # of the to be replaced file, not the dataset.
+            # TODO: PR to pyDataverse once confirmed.
+            response = self.api.replace_datafile(identifier=existing_id[0],
+                                                 filename=remote_file.name,
+                                                 json_str=datafile.json(),
+                                                 is_filepid=False)
+        else:
+            self.message(f"Uploading {remote_file} ...", type='debug')
+            response = self.api.upload_datafile(identifier=self.doi,
+                                                filename=local_file,
+                                                json_str=datafile.json())
+        response.raise_for_status()
+        uploaded_file = response.json()['data']['files'][0]
+        # update cache:
+        self.add_to_filelist(uploaded_file)
+        # remember dataverse's database id for this key
+        assigned_id = uploaded_file['dataFile']['id']
```

Comment:
This may answer my question from above (what happens to the file content of the old ID). There is no specific comment about this, but this makes me speculate whether `replace_datafile()` does not fully replace, but somehow reuses some information, but actually generates a new fileid.

If that is true, is a fileid content-specific, despite what it felt like so far?